lib/CL/devices/pthread/pthread_scheduler.c

```diff
@@ -422,6 +449,7 @@ pocl_pthread_driver_thread (void *p)
    * force a first FTZ setup */
   td->current_ftz = 213;
   td->num_threads = scheduler.num_threads;
+  td->last_cmd_ignored = NULL;
 
 #ifdef __linux__
   if (pocl_get_bool_option ("POCL_AFFINITY", 0))
```

Comment:
How does the subdevice splitting work if you don't pin the threads to cores to ensure work gets scheduled only to the cores in the sub device? Kernel's scheduler can migrate the threads to any core AFAIK and you don't have control for the subdevice compute units then? This is fuzzy.

Reply:
> How does the subdevice splitting work if you don't pin the threads to cores

Affinity has nothing to do with how it works, except it makes it easier to manually check that it works.

> Kernel's scheduler can migrate the threads to any core AFAIK

Of course, and it doesn't matter. Each pthread gets an index at creation; each subdevice gets its (core_start, core_count) values set up in clCreateSubDevices; then ``shall_we_run_this`` decides whether a particular pthread runs a command, based on its index fitting within command->device->[start, start+count]; if it does not, it ignores the command.

Reply:
OK. So it restricts the threads the commands can go into, thus setting an upper limit to the number of cores it occupies. But what stops the threads getting scheduled to the same cores? E.g. if one wants to partition an N core machine so half of the cores are dedicated to another part and another to other part, how does it work if you only restrict the threads the tasks are pushed to, but not where the threads are scheduled?

Reply:
> if you only restrict the threads the tasks are pushed to, but not where the threads are scheduled?

The tasks are not pushed to; they are pulled by each thread. If you're asking what stops threads getting scheduled to random CPU cores - nothing, except when you set POCL_AFFINITY=1 env var. But the code does not depend on that env var to work correctly, at worst you'll see worse performance b/c the threads will be randomly scheduled across CPU.

Reply:
OK, got it. Like discussed, let's leave improving this for post 1.0 (or other contributors).